testing/dsp_testing/test_beamforming.py

```diff
@@ -1,123 +1,693 @@
+                    raise Exception('Unexpected data dimensions: {}'.format(record_dict['data_descriptors']))
+
+            # STEP 2: BEAMFORM ANY UNBEAMFORMED DATA
+            if filetype != bfiq_filetype:
+                # need to beamform the data. 
+                antenna_list = []
+                #print(decimated_data.shape)
+                if data_description_list == ['num_antennas', 'num_sequences', 'num_samps']:
+                    for antenna in range(0, record_dict['data'].shape[0]):
+                        antenna_list.append(decimated_data[antenna,:])
+                    antenna_list = np.array(antenna_list)
+                elif data_description_list == ['num_sequences', 'num_antennas', 'num_samps']:
+                    for antenna in range(0, record_dict['data'].shape[1]):
+                        antenna_list.append(decimated_data[antenna,:])
+                    antenna_list = np.array(antenna_list)
+                else:
+                    raise Exception('Not sure how to beamform with the dimensions of this data: {}'.format(record_dict['data_descriptors']))
+
+                # beamform main array antennas only. 
+                main_antennas_mask = (record_dict['antennas_present'] < main_antenna_count)
+                intf_antennas_mask = (record_dict['antennas_present'] >= main_antenna_count)
+                decimated_beamformed_data = beamform(antenna_list[main_antennas_mask][:].copy(), beam_azms, freq) 
+                intf_decimated_beamformed_data = beamform(antenna_list[intf_antennas_mask][:].copy(), beam_azms, freq) 
+            else:
+                decimated_beamformed_data = decimated_data  
+                intf_decimated_beamformed_data = intf_decimated_data
+
+            sequence_filetype_dict['main_bf_data'] = decimated_beamformed_data # this has 2 dimensions: num_beams x num_samps for this sequence.
+            sequence_filetype_dict['intf_bf_data'] = intf_decimated_beamformed_data
+
+
+            # STEP 3: FIND THE PULSES IN THE DATA
+            for beamnum in range(0, sequence_filetype_dict['main_bf_data'].shape[0]):
+
+                len_of_data = sequence_filetype_dict['main_bf_data'].shape[1]
+                pulse_indices = find_pulse_indices(sequence_filetype_dict['main_bf_data'][beamnum], 0.09)
+                if len(pulse_indices) > len(pulses): # sometimes we get two samples from the same pulse.
+                    if math.fmod(len(pulse_indices), len(pulses)) == 0.0:
+                        step_size = int(len(pulse_indices)/len(pulses))
+                        pulse_indices = pulse_indices[step_size-1::step_size]
+                
+                pulse_points = [False if i not in pulse_indices else True for i in range(0,len_of_data)]
+                sequence_filetype_dict['pulse_indices'] = pulse_indices
+
+                # verify pulse indices make sense.
+                num_samples_in_tau_spacing = int(round(tau_spacing * 1.0e-6 * decimated_rate))  # us
```

Comment:
so the comment is indicating that tau_spacing is in us right? And you're converting to seconds with the 1.0e-6?

Reply:
yes, made clear